source/common/quic/envoy_quic_utils.cc

```diff
@@ -168,17 +168,17 @@ createConnectionSocket(const Network::Address::InstanceConstSharedPtr& peer_addr
                        Network::Address::InstanceConstSharedPtr& local_addr,
                        const Network::ConnectionSocket::OptionsSharedPtr& options,
                        const bool prefer_gro) {
-  if (local_addr == nullptr) {
-    local_addr = Network::Utility::getLocalAddress(peer_addr->ip()->version());
-  }
+  ASSERT(peer_addr != nullptr);
   size_t max_addresses_cache_size =
       Runtime::runtimeFeatureEnabled(
           "envoy.reloadable_features.quic_upstream_socket_use_address_cache_for_read")
           ? 4u
           : 0u;
   auto connection_socket = std::make_unique<Network::ConnectionSocketImpl>(
-      Network::Socket::Type::Datagram, local_addr, peer_addr,
-      Network::SocketCreationOptions{false, max_addresses_cache_size});
+      Network::Socket::Type::Datagram,
+      // The local address is used to retrieve the IoHandle, so it is required to not be null here.
```

Comment:
nits: how about "pass in a dumb local address if local_addr is null to pass in the socket interface used to create IoHandle"

Reply:
Reworded